### PR TITLE
Remove mkdir memoization

### PR DESCRIPTION
**Pull request by rgrinberg**
As demonstrated in #5779

### CODE REVIEW

File: test/blackbox-tests/test-cases/watching/sandbox-mkdir.t [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  $ start_dune
+
+  $ build test/target
+  Success
+  $ cat test/target
+  hello
+
+# CR-someday amokhov: Actually, it doesn't because of the broken memoization of
+# mkdirs in [build_system.ml].
+
```
Comment (snowleopard) [inline test/blackbox-tests/test-cases/watching/sandbox-mkdir.t:28]:
```suggestion
```

**Review by snowleopard — COMMENTED** [review]
Thanks for the fix! See a suggested removal of the stale CR.

**Review by snowleopard — APPROVED** [review]

File: src/dune_engine/action.ml [hunk 2]
```diff
@@ -154,16 +154,23 @@ let fold_one_step t ~init:acc ~f =
...
+   easier once [Action.t] is split off from the dune lang's action *)
 let chdirs =
   let rec loop acc t =
     let acc =
       match t with
-      | Chdir (dir, _) -> Path.Set.add acc dir
+      | Chdir (dir, _) -> (
+        match Path.as_in_build_dir dir with
+        | None ->
+          Code_error.raise "chdir ouside the build directory"
```
Comment (snowleopard) [inline src/dune_engine/action.ml:166]: Actually, could this be a user error too?

Comment (rgrinberg) [reply src/dune_engine/action.ml:166]: We have validation on the dune lang side so this case isn't possible.

Comment (snowleopard) [reply src/dune_engine/action.ml:166]: :+1:

**Pull request merged**